src/NUnitFramework/tests/Constraints/ThrowsExceptionConstraintTests.cs

```diff
@@ -65,19 +66,39 @@ public void SucceedsWithNonVoidReturningFunction()
         };
 
 #if ASYNC
+        private static async Task YieldAsync()
+        {
+#if NET40
+            await TaskEx.Yield();
+#else
+            await Task.Yield();
+#endif
+        }
+
+
         [Test]
         public static void CatchesAsyncException()
         {
             Assert.That(async () =>
             {
-#if NET40
-                await TaskEx.Yield();
-#else
-                await Task.Yield();
-#endif
+                await YieldAsync();
                 throw new Exception();
             }, Throws.Exception);
         }
+        
+        [Test]
+        public static void CatchesAsyncOfTException()
+        {
+            Assert.That(async () =>
+            {
+#pragma warning disable CS0162
+                await YieldAsync();
+                throw new Exception();
```

Comment:
Personally, I would prefer something like `if (DateTime.MinValue < DateTime.Now) throw new Exception();` or equivalent and the remove the `pragma warning`, but that is just my personal opinion (and actually I don't know how we do it in the rest of the codebase).

Reply:
Why is the `return 2` there though? It's unreachable.

Reply:
That was more just to make the compiler happy, since the specific test case was for a `Task<T>` which also throws. I've restructured it a bit now to not require the pragmas.

Is there an established way we do this in nunit?

Reply:
Now, now I see why the `return 2;` was there, so that the generic type argument would be inferred to be `Task<int>` instead of `Task`. I think the idiomatic C# would be to stop relying on inference and specify the type argument explicitly:

```cs
Assert.That<Task<int>>(async () =>
{
    await YieldAsync();
    throw new Exception();
}, Throws.Exception);
```